src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+            TxIn txIn = header.Coinstake.Inputs[0];
+            FetchCoinsResponse coins = this.PosParent.UtxoSet.FetchCoinsAsync(new[] { txIn.PrevOut.Hash }).GetAwaiter().GetResult();
+            if ((coins == null) || (coins.UnspentOutputs.Length != 1))
+                ConsensusErrors.ReadTxPrevFailed.Throw();
+
+            return coins;
+        }
+
+        /// <summary>
+        /// Gets and validates unspent outputs based of coins fetched from coin view.
+        /// </summary>
+        /// <param name="coins">The coins.</param>
+        private UnspentOutputs GetAndValidatePreviousUtxo(FetchCoinsResponse coins)
+        {
+            UnspentOutputs prevUtxo = coins.UnspentOutputs[0];
+            if (prevUtxo == null)
+            {
+                this.Logger.LogTrace("(-)[PREV_UTXO_IS_NULL]");
+                ConsensusErrors.ReadTxPrevFailed.Throw();
+            }
+
+            return prevUtxo;
+        }
+
+        /// <summary>
+        /// Checks if coinstake transaction is valid.
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckIfCoinstakeIsTrue(ProvenBlockHeader header)
+        {
+            if (header.Coinstake.IsCoinStake)
+                return;
+
+            this.Logger.LogTrace("(-)[PROVEN_HEADER_INVALID_MERKLE_PROOF_SIZE]");
+            ConsensusErrors.NonCoinstake.Throw();
+        }
+
+        /// <summary>
+        /// Checks whether header time is equal to the timestamp of the coinstake tx and if coinstake tx
+        /// timestamp is divisible by 16 (using timestamp mask).
+        /// </summary>
+        /// <param name="header">The proven block header.</param>
+        private void CheckHeaderAndCoinstakeTimes(ProvenBlockHeader header)
+        {
+            uint coinstakeTime = header.Coinstake.Time;
+            uint headerTime = header.Time; 
+
+            // Check if times are equal and coinstake tx time is divisible by 16.
+            if ((headerTime == coinstakeTime) && ((coinstakeTime & PosConsensusOptions.StakeTimestampMask) == 0))
+                return;
+
+            this.Logger.LogTrace("(-)[BAD_TIME]");
+            ConsensusErrors.StakeTimeViolation.Throw();
+        }
+
+        /// <summary>
+        /// Checks the coinstake age requirement.
+        /// </summary>
+        /// <param name="chainedHeader">The chained header.</param>
+        /// <param name="unspentOutputs">The unspent outputs.</param>
+        private void CheckCoinstakeAgeRequirement(ChainedHeader chainedHeader, UnspentOutputs unspentOutputs)
+        {
+            ChainedHeader prevChainedHeader = chainedHeader.Previous;
+
+            var options = (PosConsensusOptions)this.PosParent.Network.Consensus.Options;
+            int targetDepth = options.GetStakeMinConfirmations(prevChainedHeader.Height + 1, this.PosParent.Network) - 1;
+
+            int actualDepth = prevChainedHeader.Height - (int)unspentOutputs.Height;
```

Comment:
why unspentOutputs.Height is uint? int will be enough. 
hm. 
could you create an issue for that? or just create a small PR after this one goes in

Reply:
As per our discussion `int` and `uint` are the same size.